echemdb/data.py

```diff
@@ -1,31 +1,125 @@
 
 def collect_datapackages(data = datadir):
     # Collect all datapackage descriptors, see
     # https://specs.frictionlessdata.io/data-package/#metadata
     import os.path
     from glob import glob
-    descriptors = glob(os.path.join(data, '**', 'datapackage.json'), recursive=True)
+
+    ##NH  Here we want to build the datapackages from literature data on the fly
+    ## Note data could also be present already as datapackage.json or zip files
+    descriptors = glob(os.path.join(data, '**', '*.json'), recursive=True)
 
     # Read the package descriptors (does not read the actual data CSVs)
     from datapackage import Package
     packages = [Package(descriptor) for descriptor in descriptors]
 
     return packages
 
+
+def echemdb_id_from_descriptor(descriptor):
+    '''
+    We want to have a unique echemdb-id otherwise things look nasty
+    Let's use material-surface-uniqueid
+
+    e.g. namestr = descriptor['resources'][0]["name"] returns namestr = "briega-martos_2021_cation_XXX_p2_Fig1_Cs_2.3"
+    then we can do
+    '''
+    namestr = descriptor['resources'][0]["name"]
+    idend = hashlib.md5(namestr.encode()).hexdigest()[:7]
+    mat = property_reader(descriptor, "electrode material")
+    surf = property_reader(descriptor, "surface")
+
+    return "-".join([mat, surf, idend])
+
+
+
+
+
+
+
+
+def property_reader(descriptor, basic_property_name):
+    """
+    This function serves as a translation between basic property tags e.g.
+    ['electrode material', 'surface', 'electrolyte', 'reference', 'echemdb-id', 'path', 'relpath']
+    and how this is stored in datapackage descriptor
+
+    ['electrode material', 'surface', 'electrolyte', 'reference', 'echemdb-id', 'path', 'relpath']
+
+
+    """
+    def _finditem(obj, key):
+        if key in obj: return obj[key]
+        for k, v in obj.items():
+            if isinstance(v, dict):
+                item = _finditem(v, key)
+                if item is not None:
+                    return item
+
+
+
+    if basic_property_name == 'electrolyte':
+        els = []
+        for r in descriptor['electrochemical system']['electrolyte']['components']:
+            els.append(r["name"])
+        els.sort()
+        response = ", ".join(els)
+    elif basic_property_name == 'reference':
+        response = descriptor["source"]["doi"]
+
+    elif basic_property_name == 'echemdb-id':
+        # question: should we have an echemdb_id, i think yes
+        # I think we should have an algorithm that builds it from data in the descriptor e.g. hash
+        response = echemdb_id_from_descriptor(descriptor)
+
+
+    else:
+        basicname_toyaml = {'electrode material': ["working electrode", "material"],
+                            'surface': ["working electrode", "crystallographic orientation"]}
+
+        r1 = _finditem(descriptor, basicname_toyaml[basic_property_name][0])
+        response = _finditem(r1, basicname_toyaml[basic_property_name][1])
```

Comment:
I think this code is much easier to read if we make it explicit:
```
elif basic_property_name == "electrode material":
    material = _finditem(_finditem(descriptor, "working electrode"), "material")
    if type(material) is dict: material = material["name"]
    return material
```

But to be honest I don't understand why we cannot navigate explicitly to the right place and instead have to use this `_finditem`. We know the schema of the YAML exactly so we don't need to guess here it would seem to me.